(GH-1685) Add data type functions reference docs (#1691)
puppetlabs/bolt: documentation/bolt_types_reference.md

# Bolt data types

> This page lists the Bolt data types and describes how to use them.

Reviewer: ```suggestion
This page lists custom data types used in Bolt plans and their methods.
```

## `ApplyResult`

An [apply action](applying_manifest_blocks.md#return-value-of-apply-action)
returns an `ApplyResult`. An `ApplyResult` is part of a `ResultSet` object and
contains information about the apply action. You can access `ApplyResult`
functions with dot notation, using the syntax: `$apply_result.function`.

> For example, the following plan uses an apply action to install Docker as root,

Reviewer: At least to me this seems like it breaks the docs vs. reference distinction. I don't know if that's something formalized, but imo we should have examples and explanations in the docs proper, and this doc should assume that users already know how to use `apply` for example. Open to discussing it though.

and logs a message with any errors that occurred during execution using the
`error` function:

```puppet
plan testing::test(
  TargetSpec $targets,
) {
  $results = apply($targets, _catch_errors => true, _noop => true, _run_as => root) {
   include 'docker'
}
  $results.each |$result| {
    notice($result.error)}
}
```

The following functions are available to `ApplyResult` objects. 

| Function | Type returned | Description |

> |---|---|---|

Reviewer: omg this is so much better

# Bolt data types

> ## `ApplyResult`

Reviewer: Defer to @hestonhoffman on this, but I think we should dedup this documentation from https://puppet.com/docs/bolt/latest/writing_plans.html#handling-plan-function-results and link to it from there. Also need to update sentence here: https://puppet.com/docs/bolt/latest/applying_manifest_blocks.html#return-value-of-apply-action

Reviewer 2: I removed the bits from that doc that referenced data types and linked to this doc. Also fixed that link in the second doc

> An `ApplyResult` is returned from an [apply action](https://puppet.com/docs/bolt/latest/applying_manifest_blocks.html#)

Reviewer: I think this can be relative, so just `[apply action](applying_manifest_blocks.md#)

as part of a `ResultSet` object and contains information about the apply action.

The following functions are available to `ApplyResult` objects. They are accessed with
dot notation `$apply_result.function`.

### `action`

**Returns**

`String`: The type of result (`apply`)

### `error`

**Returns**

`Error`: An object constructed from the `_error` field of the result's value.

### `message`

**Returns**

`String`: The `_output` field of the result's value.

### `ok`

**Returns**

`Boolean`: Whether the result was successful.

### `report`

**Returns**

`Hash`: The Puppet report from the apply action.

### `target`

**Returns**

`Target`: The target the result is from.

### `to_data`

**Returns**

`Hash`: A serialized representation of `ApplyResult`.

## `Result`

> A `Result` is returned from an action executed on a `Target` as part of a `ResultSet`

Reviewer: ```suggestion
A `Result` is returned from an action executed on a single target and contains information about the execution. 
`Result`s are aggregated to make a `ResultSet` object.
```

Reviewer: I'm not sure how to say "The return from all your functions is actually going to be ResultSet which is an array of Results" here without being to wordy...but anything else feels confusing.

object and contains information about the execution.

The following functions are available to `Result` objects. They are accessed with
dot notation `$result.function`.

### `[]`

Accesses the value hash directly and returns the value for the key.

> **Note:** This function does not use dot notation and is called directly on the
  `Result`, i.e. `$result[key]`

**Returns**

`Data`: The accessed value.

### `action`

**Returns**

`String`: The type of result (`task`, `command`, etc.)

### `error`

**Returns**

`Error`: An object constructed from the `_error` field of the result's value.

### `message`

**Returns**

`String`: The `_output` field of the result's value.

### `ok`

**Returns**

`Boolean`: Whether the result was successful.

### `status`

**Returns**

`String`: Either `success` if the result was successful or `failure`.

### `target`

**Returns**

`Target`: The target the result is from.

### `value`

**Returns**

> `Hash`: The value of the result.

Reviewer: This doesn't feel helpful. Maybe "The result of executing the action on the target"

## `ResultSet`

A `ResultSet` is returned from an execution function and contains one or more `Result`

> objects. An [apply action](https://puppet.com/docs/bolt/latest/applying_manifest_blocks.html#)

Reviewer: Relative link

`ResultSet`: The set of filtered results.

### `find(String $target_name)`

**Parameters**

- `String` target_name

  The string name of the target to retrive a result for.

**Returns**

`Variant[Result, ApplyResult]`: The target's result.

### `first`

**Returns**

`Variant[Result, ApplyResult]`: The first result in the set. Useful for unwrapping single results.

### `names`

**Returns**

`Array[String]`: The names of all targets that have results in the set.

### `ok`

**Returns**

`Boolean`: Whether all results were successful. Equivalent to `$result_set.error_set.empty`.

### `ok_set`

**Returns**

`ResultSet`: The set of successful results.

### `results`

**Returns**

`Array[Variant[Result, ApplyResult]]`: All results in the set.

### `targets`

**Returns**

`Array[Target]`: The list of targets that have results in the set.

### `to_data`

**Returns**

`Array[Hash`: An array of serialized representations of each result in the set.

## `Target`

The `Target` object represents a target and its specific connection options.

The following functions are available to `ResultSet` objects. They are accessed with
dot notation `$result_set.function`.

### `config`

**Returns**

> `Hash[String, Data]`: The target-level configuration.

Reviewer: I think including a line about how this is only data from the inventory, though I think it's all inventory data that applies to a target (not just defined specifically for that target)

Reviewer: This in addition to the note below.

> **Note:** This function does not return default configuration values or
  configuration set in a `bolt.yaml` file.

### `facts`

**Returns**

`Hash[String, Data]`: The target's facts.

> **Note::** This function does not lookup facts for a target and only returns
  the facts specified in an `inventory.yaml` file or set on a target during a
  plan run.

### `features`

**Returns**

`Array[String]`: The target's features.

### `host`

**Returns**

`String`: The target's hostname.

### `name`

**Returns**

`String`: The target's human-readable name, or its URI if a name was not given.

### `password`

**Returns**

`String`: The password to use when connecting to the target.

### `plugin_hooks`

**Returns**

`Hash[String, Data]`: The target's 
[`plugin_hooks` configuration options](https://puppet.com/docs/bolt/latest/bolt_configuration_reference.html#plugin-hooks-configuration-options)

### `port`

**Returns**

`Integer`: The target's connection port.

### `protocol`

**Returns**

> `String`: The protocol used to connect to the target.

Reviewer: Include that this is the same as the transport except for 'remote' transport

### `safe_name`

**Returns**

`String`: The target's safe name. Equivalent to `name` if a name was given, or
the target's `uri` with any password omitted.

### `target_alias`

**Returns**

`Variant[String, Array[String]]`: The target's aliases.

### `uri`

**Returns**

`String`: The target's URI.

### `user`

**Returns**

> `String`: The user to run as on the target.

Reviewer: ```suggestion
`String`: The user to connect to the target.
```

> ### `vars`

Reviewer: https://puppet.com/docs/puppet/latest/lang_data.html I don't see the `Error` datatype documented in Puppet, but we should probably find a way to document it and include it since we reference it. Ideally adding it to the Puppet docs and linking would be the way to go, but if that's going to take a while including here might not be a bad idea?

Reviewer 2: I'll open a ticket for this bit and add it to the Puppet docs
